### DISCUSSION TITLE
Best way for large scale batch inference MMSeg TensorRT model?

### DISCUSSION THREAD

**Discussion by davodogster** [open-mmlab/mmdeploy#1457] [General]
Hello MMDeploy community,
I trained an MMSeg model and converted it to ONNX and then TRT and I was able to do successful inference on a single image using mmdeploy.apis.inference_model. 

Is it possible to do large scale batch inference with an MMSeg TensorRT model? I am using a custom Test Time augmentation function so I would like the batch size to be a minimum of 8. I have 100,000s of images to run through the pipeline on a GPU (and I plan to run it on AWS).

It appears there are at least 3 options (for batch=1?) like Model Converter, SDK, Python API. Thanks, Sam

@RunningLeon

**Comment by davodogster** [comment]
@lvhan028 I tried this in SDK Segmentation with my onnx model but it failed.
![image](https://user-images.githubusercontent.com/46079516/205208139-3bd609f0-7a29-4a36-b303-4d3502719ccd.png)

SDK inference looks very complex compared to inference_model in python

**Reply by lvhan028** [reply]
According to the log, loading SDK model failed. 
Can you run `tree ../mmdeploy/workdir` and paste the result here?
Also could you share the command which is used to convert the MMSeg model?

**Reply by davodogster** [reply]
Hi @lvhan028 

Here is the tree command:
![image](https://user-images.githubusercontent.com/46079516/206332983-ae6f2fc4-0a98-4621-9161-683ed4598305.png)

Command to convert the mmseg model to onnx is here:
`python ~/workspace/mmdeploy/tools/deploy.py \
  ~/workspace/mmdeploy/configs/mmseg/segmentation_onnxruntime_dynamic.py \
  /workspace/mounted_dir/output/FM_binary-mmseg-pspnet50/fcn_unet.py \
  /workspace/mounted_dir/output/FM_binary-mmseg-pspnet50/latest.pth \
  /workspace/mounted_dir/data/FM1-indexed-split/test/AT24_1000_3948__7.jpg \
  --test-img /workspace/mounted_dir/data/FM1-indexed-split/test/AT24_1000_3948__7.jpg`

Thanks, Sam

**Reply by lvhan028** [reply]
please append `--dump-info` in the command. It will generate meta info used by mmdeploy SDK

**Reply by davodogster** [reply]
@lvhan028 thanks for your reply - good to know about --dump-info.

I ran it with dump info and copied the files into /workdir and have rerun the code. Here is the output:

![image](https://user-images.githubusercontent.com/46079516/206347918-f30dce23-fff7-43ed-8d6e-bf4f89bb408f.png)

Can I fix this? Cheers, Sam

To generate the .onnx file I used the cpu docker image docker run -it --rm -v `pwd`:/workspace/mounted_dir mmdeploy:master-cpu

**Reply by lvhan028** [reply]
It seems that `onnxruntime` backend is not built. What's your command to build SDK?
`-DMMDEPLOY_TARGET_BACKENDS="ort;trt"` should be passed to invoke both onnxruntime backend and tensorrt backend

**Comment by davodogster** [comment]
Thanks @lvhan028 . My other question is: Can TensorRT SDK be used inside the prebuilt MMdeploy:GPU docker image? Is it possible and easy to configure?

**Reply by lvhan028** [reply]
Yes. It can. 
You can refer to https://github.com/open-mmlab/mmdeploy/blob/master/docs/en/01-how-to-build/build_from_docker.md

**Comment by davodogster** [comment]
@lvhan028 Thanks! Now the Segmenter is very giving different results to TRTWrapper. Both run inside the same Python environment.

Segmenter only takes NP arrays. The TRTWrapper I am feeding a CUDA Torch Tensor

![image](https://user-images.githubusercontent.com/46079516/208816392-5d6316f7-11b9-42e9-ac06-a6ea34b852ad.png)

![image](https://user-images.githubusercontent.com/46079516/208816646-889c6696-f489-4cbd-9363-4716106aed8b.png)

And is it okay to ignore: ?

[12/22/2022-14:43:37] [TRT] [W] TensorRT was linked against cuBLAS/cuBLAS LT 11.6.5 but loaded cuBLAS/cuBLAS LT 111.0.3

**Reply by lvhan028** [reply]
Yes. Warnings can be ignored.
But I don't get your question. Do you mean that the result inferred by `mmdepoy_python.Segmentor` is different from that inferred by TRTWrapper?

**Comment by davodogster** [comment]
Yes that's what I mean - the results are very different. It should be predicting 100K+pixels for 0 and 1.

I have visualised the image and the tensor and they look the same. Does the TRTWrapper automatically preprocess the image? I assume it does?Thanks, Sam
________________________________

**Reply by davodogster** [reply]
@lvhan028 Hi do you know why the results could be very different using the same .engine model? They give very different results for the same input tensor:

![image](https://user-images.githubusercontent.com/46079516/211457169-f0bc98a2-f31c-49ef-b7a9-8215af172958.png)


![image](https://user-images.githubusercontent.com/46079516/211456453-2ab5bf42-867a-42da-b9df-f70a76a19b1b.png)

**Reply by davodogster** [reply]
@lvhan028 does the TRTWrapper model automatically do the image normalization? Or do I need to code it in?

![image](https://user-images.githubusercontent.com/46079516/211458228-ed073978-787d-4fd3-9e91-2d56bac7caf6.png)

**Reply by RunningLeon** [reply]
@davodogster Yes. The normalization will be done in sdk model automatically.

**Reply by davodogster** [reply]
@RunningLeon Ok thats great - so why do you think I'm getting different results for TRTWrapper vs Segmentor when I run them on the same image inside python?

**Reply by davodogster** [reply]
@lvhan028 @RunningLeon Segmentor is giving the correct results (predicting many 0s and 1s) but TRTWrapper model is not giving the right results (predicting mostly 0's).

**Reply by RunningLeon** [reply]
@davodogster Hi, could you provide a small model and some sample codes to reproduce the issue?

**Reply by davodogster** [reply]
@RunningLeon Sure thing I have just emailed you at mnsheng @ yeah.net with the model and code etc. Please let me know if you didn't receive the email. Sam

**Reply by RunningLeon** [reply]
@davodogster Yes. I have received your email and would look through it later. BTW, could you run `python tools/check_env.py` and post env info in here?

**Reply by davodogster** [reply]
@RunningLeon mmdeploy_trt environment check_env:

2023-01-12 15:11:35,900 - mmdeploy - INFO - **********Environmental information**********
2023-01-12 15:11:36,223 - mmdeploy - INFO - sys.platform: linux
2023-01-12 15:11:36,224 - mmdeploy - INFO - Python: 3.9.15 (main, Nov  4 2022, 16:13:54) [GCC 11.2.0]
2023-01-12 15:11:36,224 - mmdeploy - INFO - CUDA available: True
2023-01-12 15:11:36,224 - mmdeploy - INFO - GPU 0: NVIDIA GeForce RTX 3090
2023-01-12 15:11:36,224 - mmdeploy - INFO - CUDA_HOME: /usr/local/cuda
2023-01-12 15:11:36,224 - mmdeploy - INFO - NVCC: Cuda compilation tools, release 11.6, V11.6.112
2023-01-12 15:11:36,224 - mmdeploy - INFO - GCC: gcc (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0
2023-01-12 15:11:36,224 - mmdeploy - INFO - PyTorch: 1.13.0+cu117
2023-01-12 15:11:36,224 - mmdeploy - INFO - PyTorch compiling details: PyTorch built with:
  - GCC 9.3
  - C++ Version: 201402
  - Intel(R) Math Kernel Library Version 2020.0.0 Product Build 20191122 for Intel(R) 64 architecture applications
  - Intel(R) MKL-DNN v2.6.0 (Git Hash 52b5f107dd9cf10910aaa19cb47f3abf9b349815)
  - OpenMP 201511 (a.k.a. OpenMP 4.5)
  - LAPACK is enabled (usually provided by MKL)
  - NNPACK is enabled
  - CPU capability usage: AVX2
  - CUDA Runtime 11.7
  - NVCC architecture flags: -gencode;arch=compute_37,code=sm_37;-gencode;arch=compute_50,code=sm_50;-gencode;arch=compute_60,code=sm_60;-gencode;arch=compute_70,code=sm_70;-gencode;arch=compute_75,code=sm_75;-gencode;arch=compute_80,code=sm_80;-gencode;arch=compute_86,code=sm_86
  - CuDNN 8.2.1  (built against CUDA 11.3)
    - Built with CuDNN 8.5
  - Magma 2.6.1
  - Build settings: BLAS_INFO=mkl, BUILD_TYPE=Release, CUDA_VERSION=11.7, CUDNN_VERSION=8.5.0, CXX_COMPILER=/opt/rh/devtoolset-9/root/usr/bin/c++, CXX_FLAGS= -fabi-version=11 -Wno-deprecated -fvisibility-inlines-hidden -DUSE_PTHREADPOOL -fopenmp -DNDEBUG -DUSE_KINETO -DUSE_FBGEMM -DUSE_QNNPACK -DUSE_PYTORCH_QNNPACK -DUSE_XNNPACK -DSYMBOLICATE_MOBILE_DEBUG_HANDLE -DEDGE_PROFILER_USE_KINETO -O2 -fPIC -Wno-narrowing -Wall -Wextra -Werror=return-type -Werror=non-virtual-dtor -Wno-missing-field-initializers -Wno-type-limits -Wno-array-bounds -Wno-unknown-pragmas -Wunused-local-typedefs -Wno-unused-parameter -Wno-unused-function -Wno-unused-result -Wno-strict-overflow -Wno-strict-aliasing -Wno-error=deprecated-declarations -Wno-stringop-overflow -Wno-psabi -Wno-error=pedantic -Wno-error=redundant-decls -Wno-error=old-style-cast -fdiagnostics-color=always -faligned-new -Wno-unused-but-set-variable -Wno-maybe-uninitialized -fno-math-errno -fno-trapping-math -Werror=format -Werror=cast-function-type -Wno-stringop-overflow, LAPACK_INFO=mkl, PERF_WITH_AVX=1, PERF_WITH_AVX2=1, PERF_WITH_AVX512=1, TORCH_VERSION=1.13.0, USE_CUDA=ON, USE_CUDNN=ON, USE_EXCEPTION_PTR=1, USE_GFLAGS=OFF, USE_GLOG=OFF, USE_MKL=ON, USE_MKLDNN=ON, USE_MPI=OFF, USE_NCCL=ON, USE_NNPACK=ON, USE_OPENMP=ON, USE_ROCM=OFF,

2023-01-12 15:11:36,224 - mmdeploy - INFO - TorchVision: 0.14.0+cu117
2023-01-12 15:11:36,224 - mmdeploy - INFO - OpenCV: 4.6.0
2023-01-12 15:11:36,224 - mmdeploy - INFO - MMCV: 1.7.0
2023-01-12 15:11:36,224 - mmdeploy - INFO - MMCV Compiler: GCC 9.3
2023-01-12 15:11:36,225 - mmdeploy - INFO - MMCV CUDA Compiler: 11.7
2023-01-12 15:11:36,225 - mmdeploy - INFO - MMDeploy: 0.10.0+de96f51
2023-01-12 15:11:36,225 - mmdeploy - INFO -

2023-01-12 15:11:36,225 - mmdeploy - INFO - **********Backend information**********
2023-01-12 15:11:36,619 - mmdeploy - INFO - onnxruntime: None   ops_is_avaliable : False
2023-01-12 15:11:36,642 - mmdeploy - INFO - tensorrt: 8.2.3.0   ops_is_avaliable : True
2023-01-12 15:11:36,655 - mmdeploy - INFO - ncnn: None  ops_is_avaliable : False
2023-01-12 15:11:36,656 - mmdeploy - INFO - pplnn_is_avaliable: False
2023-01-12 15:11:36,657 - mmdeploy - INFO - openvino_is_avaliable: False
2023-01-12 15:11:36,670 - mmdeploy - INFO - snpe_is_available: False
2023-01-12 15:11:36,671 - mmdeploy - INFO - ascend_is_available: False
2023-01-12 15:11:36,673 - mmdeploy - INFO - coreml_is_available: False
2023-01-12 15:11:36,673 - mmdeploy - INFO -

2023-01-12 15:11:36,673 - mmdeploy - INFO - **********Codebase information**********
2023-01-12 15:11:36,675 - mmdeploy - INFO - mmdet:      None
2023-01-12 15:11:36,675 - mmdeploy - INFO - mmseg:      0.29.1
2023-01-12 15:11:36,675 - mmdeploy - INFO - mmcls:      0.24.1
2023-01-12 15:11:36,675 - mmdeploy - INFO - mmocr:      None
2023-01-12 15:11:36,675 - mmdeploy - INFO - mmedit:     None
2023-01-12 15:11:36,675 - mmdeploy - INFO - mmdet3d:    None
2023-01-12 15:11:36,675 - mmdeploy - INFO - mmpose:     None
2023-01-12 15:11:36,675 - mmdeploy - INFO - mmrotate:   None

**Reply by RunningLeon** [reply]
@davodogster Hi, I've looked your sample code and tried on my machine with official model config and ckpt.


1. If you want to compare results from `TRT engine` and `mmdeploy sdk model` in python, you'd better to use `from mmdeploy.apis import visualize_model` and [sdk python demo](https://github.com/open-mmlab/mmdeploy/blob/v0.10.0/demo/python/image_segmentation.py) and compare the visualized result images.
2. `TRTWrapper` cannot produce same outputs as `from mmdeploy_python import Segmentor` because `TRTWrapper` does not have some post-processing like resize segment output to original shape of input image.
3. If you still want to compare output tensors directly, you could refer to following sample code(you cannot get exactly same outputs, but they are close enough).

```
from mmdeploy_python import Segmentor
from mmdeploy.apis import build_task_processor

import cv2
import torch
import numpy as np
import mmcv

# img_path = "/home/scion.local/davidsos/Documents/mmdeploy_trt/mmdeploy/demo/BF38_1000_4541__5.jpg"
# sdk_model_path = "/home/scion.local/davidsos/Documents/mmdeploy_trt/mmdeploy/workdir_new"

img_path = "../mmsegmentation/demo/demo.png"
sdk_model_path = "/root/workspace/mmdeploy_regression_working_dir/mmseg/torch1.10.0/mmseg/fcn/tensorrt/dynamic/fp16/fcn_r50-d8_512x1024_40k_cityscapes_20200604_192608-efe53f0d"
model_cfg_path = '../mmsegmentation/configs/fcn/fcn_r50-d8_512x1024_40k_cityscapes.py'
deploy_cfg_path = './configs/mmseg/segmentation_tensorrt-fp16_dynamic-512x1024-2048x2048.py'

model_cfg = mmcv.Config.fromfile(model_cfg_path)
deploy_cfg = mmcv.Config.fromfile(deploy_cfg_path)
trt_engine_path = sdk_model_path + '/end2end.engine'
device = 'cuda:0'
task_processor = build_task_processor(model_cfg, deploy_cfg, device)
model_inputs, input_tensor = task_processor.create_input(img_path, None)

trt_model = task_processor.init_backend_model([trt_engine_path])
trt_output = task_processor.run_inference(trt_model, model_inputs)[0]

segmentor = Segmentor(sdk_model_path, "cuda", 0)
image = cv2.imread(img_path)
seg_output = segmentor(image)
print(np.unique(seg_output, return_counts=True))
print(np.unique(trt_output, return_counts=True))
# outputs shows they are close.
"""

(array([ 0,  1,  2,  5,  6,  7,  8, 10, 11, 12, 13, 14, 15, 16, 18],
      dtype=int32), array([214885,  36814, 168822,  11975,   3801,   2956,   5757,  37408,
         4505,   1755,   8498,     47,    227,    105,  26733]))
(array([ 0,  1,  2,  5,  6,  7,  8, 10, 11, 12, 13, 14, 15, 16, 18]), array([214880,  36875, 168744,  11936,   3807,   2967,   5750,  37412,
         4489,   1742,   8517,     38,    223,    114,  26794]))
"""
```

**Reply by davodogster** [reply]
Hi @RunningLeon thanks for taking the time to investigate that! I have just ran it on my machine with my demo image and you are right that they give 1. similar predictions and importantly 2. good predictions.

Ultimately, all I want to do is very fast large scale inference e.e 500,000 + images, hence why I am using TRT instead of Torch for inference. I also assume that using batch size > 1 will speed up my large scale inference. But it appears `task_processor` and `Segmetor` do not allow for a batch of input tensors/arrays? What would you recommend for MMDeploy TRT .engine fast large scale inference? It must be a standard use case? We also have our own inference pipeline with a test time augmentation function that we want to wrap around MMDeploy TRT inference.

Thanks, Sam

1. If you want to compare results from TRT engine and mmdeploy sdk model in python, you'd better to use from mmdeploy.apis import visualize_model and [sdk python demo](https://github.com/open-mmlab/mmdeploy/blob/v0.10.0/demo/python/image_segmentation.py) and compare the visualized result images. **I'm not too interested in comparisons, I'm interested in fast large scale accurate inference, where I can include my own test time augmentation code.**
2. TRTWrapper cannot produce same outputs as from mmdeploy_python import Segmentor because TRTWrapper does not have some post-processing like resize segment output to original shape of input image. **Ok that is good to know. The results I were getting were extremely different so I would be surprised if postprocessing can give the correct results.**
3. If you still want to compare output tensors directly, you could refer to following sample code(you cannot get exactly same outputs, but they are close enough). Thanks I have done that.

![image](https://user-images.githubusercontent.com/46079516/212192251-27ef5d3c-f604-48fc-9459-5ee40bde0fa3.png)

**Reply by RunningLeon** [reply]
@davodogster Hi, actually, `Segmentor` suppports batch inference. You could try `mmdeploy==0.12.0` and recreate your sdk model with `tools/deploy.py` and then do batch inference.
```
image = cv2.imread('xxx')
segmentor = Segmentor(
        model_path=sdk_model_path, device_name='cuda', device_id=0)
image_list = [image, image, image]
results = segmentor.batch(image_list)
```

**Reply by davodogster** [reply]
@RunningLeon Thanks I'll try that on Monday with mmdeploy 0.12.0 assuming that 0.10 will not do batch inference. I'm currently using MMDeploy GPU container for the model conversion and a separate MMdeploy TRT local conda environment for the inference.

**Reply by davodogster** [reply]
EDITED: @RunningLeon Thanks I have installed 0.12.0 docker TRT GPU. I received this error: Do you know how I can solve it?

`python mmdeploy/tools/deploy.py \
    mmdeploy/configs/mmseg/segmentation_tensorrt-fp16_dynamic-512x1024-2048x2048.py \
    /workspace/mounted_dir/output/FM_binary-mmseg-pspnet50/fcn_unet.py \
    /workspace/mounted_dir/output/FM_binary-mmseg-pspnet50/latest.pth \
    /workspace/mounted_dir/data/FM1-indexed-split/test/AT24_1000_3948__7.jpg \
    --work-dir /workspace/mounted_dir/FM_Unet_TRT_012_BS8_v2 \
    --device cuda \
    --dump-info`

```
python
from mmdeploy_python import Segmentor
from mmdeploy.apis import build_task_processor
import cv2
import torch
import numpy as np
import mmcv
sdk_model_path = "/workspace/mounted_dir/FM_Unet_TRT_012_BS8/"
segmentor = Segmentor(sdk_model_path, "cuda", 0)
img_path = "/workspace/mounted_dir/BF38_1000_4541__5.jpg"
image = cv2.imread(img_path)
image=image.astype('float32')
image_list = [image, image, image,image, image, image,image, image]
results = segmentor.batch(image_list)
```

  "min_shape": [
      1,
      3,
      512,
      992
  ],
  "opt_shape": [
      4,3,1024,1024 ],
  "max_shape": [
      8,3,1024,1024
  ]

![image](https://user-images.githubusercontent.com/46079516/213072239-d6f3dbc1-411b-4bc4-b924-d2e2c4d3996c.png)
![image](https://user-images.githubusercontent.com/46079516/213082649-4eee8116-d684-4372-b0be-dec1ce26f7d7.png)

check_env:

`2023-01-18 02:57:48,726 - mmdeploy - INFO - PyTorch: 1.10.0
2023-01-18 02:57:48,726 - mmdeploy - INFO - PyTorch compiling details: PyTorch built with:
  - GCC 7.3
  - C++ Version: 201402
  - Intel(R) Math Kernel Library Version 2020.0.2 Product Build 20200624 for Intel(R) 64 architecture applications
  - Intel(R) MKL-DNN v2.2.3 (Git Hash 7336ca9f055cf1bfa13efb658fe15dc9b41f0740)
  - OpenMP 201511 (a.k.a. OpenMP 4.5)
  - LAPACK is enabled (usually provided by MKL)
  - NNPACK is enabled
  - CPU capability usage: AVX512
  - CUDA Runtime 11.3
  - NVCC architecture flags: -gencode;arch=compute_37,code=sm_37;-gencode;arch=compute_50,code=sm_50;-gencode;arch=compute_60,code=sm_60;-gencode;arch=compute_61,code=sm_61;-gencode;arch=compute_70,code=sm_70;-gencode;arch=compute_75,code=sm_75;-gencode;arch=compute_80,code=sm_80;-gencode;arch=compute_86,code=sm_86;-gencode;arch=compute_37,code=compute_37
  - CuDNN 8.2
  - Magma 2.5.2
  - Build settings: BLAS_INFO=mkl, BUILD_TYPE=Release, CUDA_VERSION=11.3, CUDNN_VERSION=8.2.0, CXX_COMPILER=/opt/rh/devtoolset-7/root/usr/bin/c++, CXX_FLAGS= -Wno-deprecated -fvisibility-inlines-hidden -DUSE_PTHREADPOOL -fopenmp -DNDEBUG -DUSE_KINETO -DUSE_FBGEMM -DUSE_QNNPACK -DUSE_PYTORCH_QNNPACK -DUSE_XNNPACK -DSYMBOLICATE_MOBILE_DEBUG_HANDLE -DEDGE_PROFILER_USE_KINETO -O2 -fPIC -Wno-narrowing -Wall -Wextra -Werror=return-type -Wno-missing-field-initializers -Wno-type-limits -Wno-array-bounds -Wno-unknown-pragmas -Wno-sign-compare -Wno-unused-parameter -Wno-unused-variable -Wno-unused-function -Wno-unused-result -Wno-unused-local-typedefs -Wno-strict-overflow -Wno-strict-aliasing -Wno-error=deprecated-declarations -Wno-stringop-overflow -Wno-psabi -Wno-error=pedantic -Wno-error=redundant-decls -Wno-error=old-style-cast -fdiagnostics-color=always -faligned-new -Wno-unused-but-set-variable -Wno-maybe-uninitialized -fno-math-errno -fno-trapping-math -Werror=format -Wno-stringop-overflow, LAPACK_INFO=mkl, PERF_WITH_AVX=1, PERF_WITH_AVX2=1, PERF_WITH_AVX512=1, TORCH_VERSION=1.10.0, USE_CUDA=ON, USE_CUDNN=ON, USE_EXCEPTION_PTR=1, USE_GFLAGS=OFF, USE_GLOG=OFF, USE_MKL=ON, USE_MKLDNN=ON, USE_MPI=OFF, USE_NCCL=ON, USE_NNPACK=ON, USE_OPENMP=ON,

2023-01-18 02:57:48,726 - mmdeploy - INFO - TorchVision: 0.11.0
2023-01-18 02:57:48,727 - mmdeploy - INFO - OpenCV: 4.7.0
2023-01-18 02:57:48,727 - mmdeploy - INFO - MMCV: 1.5.3
2023-01-18 02:57:48,727 - mmdeploy - INFO - MMCV Compiler: GCC 7.3
2023-01-18 02:57:48,727 - mmdeploy - INFO - MMCV CUDA Compiler: 11.3
2023-01-18 02:57:48,727 - mmdeploy - INFO - MMDeploy: 0.12.0+9d3b494
2023-01-18 02:57:48,727 - mmdeploy - INFO -

2023-01-18 02:57:48,727 - mmdeploy - INFO - **********Backend information**********
2023-01-18 02:57:48,783 - mmdeploy - INFO - tensorrt:   8.2.4.2
2023-01-18 02:57:48,783 - mmdeploy - INFO - tensorrt custom ops:        Available
2023-01-18 02:57:48,807 - mmdeploy - INFO - ONNXRuntime:        None
2023-01-18 02:57:48,808 - mmdeploy - INFO - ONNXRuntime-gpu:    1.8.1
2023-01-18 02:57:48,808 - mmdeploy - INFO - ONNXRuntime custom ops:     Available
2023-01-18 02:57:48,809 - mmdeploy - INFO - pplnn:      None
2023-01-18 02:57:48,812 - mmdeploy - INFO - ncnn:       None
2023-01-18 02:57:48,814 - mmdeploy - INFO - snpe:       None
2023-01-18 02:57:48,815 - mmdeploy - INFO - openvino:   None
2023-01-18 02:57:48,817 - mmdeploy - INFO - torchscript:        1.10.0
2023-01-18 02:57:48,818 - mmdeploy - INFO - torchscript custom ops:     NotAvailable
2023-01-18 02:57:48,844 - mmdeploy - INFO - rknn-toolkit:       None
2023-01-18 02:57:48,844 - mmdeploy - INFO - rknn2-toolkit:      None
2023-01-18 02:57:48,845 - mmdeploy - INFO - ascend:     None
2023-01-18 02:57:48,847 - mmdeploy - INFO - coreml:     None
2023-01-18 02:57:48,848 - mmdeploy - INFO - tvm:        None
2023-01-18 02:57:48,848 - mmdeploy - INFO -

2023-01-18 02:57:48,848 - mmdeploy - INFO - **********Codebase information**********
2023-01-18 02:57:48,850 - mmdeploy - INFO - mmdet:      None
2023-01-18 02:57:48,850 - mmdeploy - INFO - mmseg:      0.30.0
2023-01-18 02:57:48,850 - mmdeploy - INFO - mmcls:      0.25.0
2023-01-18 02:57:48,850 - mmdeploy - INFO - mmocr:      None
2023-01-18 02:57:48,850 - mmdeploy - INFO - mmedit:     None
2023-01-18 02:57:48,850 - mmdeploy - INFO - mmdet3d:    None
2023-01-18 02:57:48,850 - mmdeploy - INFO - mmpose:     None
2023-01-18 02:57:48,850 - mmdeploy - INFO - mmrotate:   None
2023-01-18 02:57:48,850 - mmdeploy - INFO - mmaction:   None`

**Reply by RunningLeon** [reply]
@davodogster Hi, sorry for the trouble. Could try to change this line 
https://github.com/open-mmlab/mmdeploy/blob/c3986cebe8e3d5125e1b70fdde03bbbecfd0f6b1/csrc/mmdeploy/utils/opencv/opencv_utils.cpp#L129 to `cv::resize(src, dst, dst.size(), 0, 0, method);` and rebuild the mmdeploy from source? You could build in your docker container according to https://github.com/open-mmlab/mmdeploy/blob/v0.12.0/docs/en/01-how-to-build/build_from_source.md

**Reply by davodogster** [reply]
@RunningLeon Oh that seems inconvenient? Do you mean build it inside my docker container which already has everything built and installed?
![image](https://user-images.githubusercontent.com/46079516/213310908-bdcccc0f-199e-4e79-b746-45fe856792df.png)

How would I rebuild it? Do I need to run all the code here for linux? Or just certain parts of it? https://github.com/open-mmlab/mmdeploy/blob/v0.12.0/docs/en/01-how-to-build/linux-x86_64.md

**Reply by davodogster** [reply]
I tried inside MMdeploy DOcker and it failed

![image](https://user-images.githubusercontent.com/46079516/213314577-408faa30-a9b1-4d6a-8a8f-3710f13687c8.png)

Remember my goal is just large scale GPU TRT inference - I will use whatever works best that mmdeploy offers. Thanks

**Reply by RunningLeon** [reply]
As it shows in the [doc](https://github.com/open-mmlab/mmdeploy/blob/v0.12.0/docs/en/01-how-to-build/linux-x86_64.md#install-toolchains), you have to update g++ and gcc to >=7. You could check current version by `g++ --version & gcc --version`. If it satisfies, you could change `g++-7` to `g++` in building script.

**Reply by davodogster** [reply]
@RunningLeon Thanks, the mmdep docker container contains:

![image](https://user-images.githubusercontent.com/46079516/213335214-dfc9312b-232b-4722-b3ff-2f4ee0b35db0.png)

It's >= 7 right?

Which building script do you mean for me to edit?
Sorry I'm not very familiar with cmake

Do I simply change this part?
![image](https://user-images.githubusercontent.com/46079516/213338696-769849ef-1799-44d8-9303-2435770e306c.png)

**Reply by RunningLeon** [reply]
Yes

**Reply by davodogster** [reply]
@RunningLeon Thanks that part seems to have worked!

![image](https://user-images.githubusercontent.com/46079516/213339378-0a031336-037d-4bbb-9750-e08dd703300e.png)

Ill try next steps now to rebuild SDK and then run the Segmentor

**Reply by davodogster** [reply]
@RunningLeon Thanks! batch inference is working properly now.

Here is the speed for batch size=8, image shape = 992,992,3 for FCN_Unet model
![image](https://user-images.githubusercontent.com/46079516/213346389-2e42de2b-a2da-49ee-921f-de4ce32b7daf.png)

1. Do you think it's significantly faster than using inference with the .pth model (are there any mmdeploy speed benchmarking tests available) ?
2. Can I pass and return torch cuda tensors to the segmentor() function instead of numpy array? If it will speed up inference. My wrapper function expects tensors as input.

**Reply by RunningLeon** [reply]
Hi,
1. normally tensorrt should be faster than pytorch, but you still have to test on your device.
2. mmdeploy does not provide this feature of `Tensor` input to segmentor in Python API.

**Reply by davodogster** [reply]
> @davodogster Hi, sorry for the trouble. Could try to change this line
> 
> https://github.com/open-mmlab/mmdeploy/blob/c3986cebe8e3d5125e1b70fdde03bbbecfd0f6b1/csrc/mmdeploy/utils/opencv/opencv_utils.cpp#L129
> 
> to `cv::resize(src, dst, dst.size(), 0, 0, method);` and rebuild the mmdeploy from source? You could build in your docker container according to https://github.com/open-mmlab/mmdeploy/blob/v0.12.0/docs/en/01-how-to-build/build_from_source.md

Hi @RunningLeon Will there be a future version of the repo (and docker image) that fixes this cv error which has to be manually changed?

**Reply by RunningLeon** [reply]
@davodogster Yes, it will be fixed in this PR #1625 https://github.com/open-mmlab/mmdeploy/pull/1625/files#diff-15b6fdbc4249496c473fce836516ea8bdd71bc3e9a9432f4f7dc6454362d61c6
